### PR TITLE
LinkedList enumerator

### LINKED ISSUE #21: LinkedItem enumerator
- [ ] Add `enumerate(key: &Key)` to be able to iterate all items for a given key. Similar to [`LinkedStorageMap::enumerate`](https://github.com/paritytech/substrate/blob/a86bb37b7c9eb7d583e5554fc0790050beaa14d5/srml/support/src/storage/generator/linked_map.rs#L338)
- [ ] Add `take_all(key: &Key)` to return an enumerator that iterates all items for a given key and remove all items at same time. This enumerator should implement `Drop` to delete all un-iterated  items.

### PR DESCRIPTION
**Pull request by shaopengw**
Closes #21

### CODE REVIEW

File: utilities/src/linked_item.rs [hunk 1]
```diff
@@ -74,8 +87,68 @@ where
...
+// 	Storage: StorageMap<(Key, Option<Value>), LinkedItem<Value>, Query = Option<LinkedItem<Value>>>,
+// {
+// 	fn drop(&mut self) {
+// 		if !self.should_take {
+// 			return
+// 		}
+
+// 		while self.next() != None {}
+// 	}
+// }
```
Comment (shaopengw) [inline utilities/src/linked_item.rs]: This straightforward implementation doesn't work:
> error[E0366]: Implementations of Drop cannot be specialized

**Comment by shaopengw** [comment]
`Drop` could not be implemented in this straightforward way.

> error[E0366]: Implementations of Drop cannot be specialized

https://github.com/laminar-protocol/open-runtime-module-library/blob/3ddd63ec16dade262eed7056934c7ad1a9ec8f64/utilities/src/linked_item.rs#L136-L150

File: utilities/src/linked_item.rs [hunk 2]
```diff
@@ -8,9 +9,15 @@ pub struct LinkedItem<Item> {
 	pub next: Option<Item>,
 }
 
+impl<Item> Default for LinkedItem<Item> {
```
Comment (xlc) [inline utilities/src/linked_item.rs]: can you derive Default?

File: utilities/src/linked_item.rs [hunk 3]
```diff
@@ -20,12 +27,18 @@ where
...
 
-	fn write_head(account: &Key, item: LinkedItem<Value>) {
-		Self::write(account, None, item);
+	fn write_head(key: &Key, item: LinkedItem<Value>) {
+		Self::write(key, None, item);
 	}
 
 	fn read(key: &Key, value: Option<Value>) -> LinkedItem<Value> {
-		Storage::get(&(key.clone(), value)).unwrap_or_else(|| LinkedItem { prev: None, next: None })
+		Storage::get(&(key.clone(), value)).unwrap_or_else(|| Default::default())
```
Comment (xlc) [inline utilities/src/linked_item.rs]:
```suggestion
		Storage::get((&key, value)).unwrap_or_else(Default::default)
```

File: utilities/src/linked_item.rs (hunk 2 above)
Comment (shaopengw) [reply utilities/src/linked_item.rs]: rustc requires `Value: core::default::Default` bound to use `Default::default` in LinkedList impl, might makes `Value` kind of restricted?

Comment (xlc) [reply utilities/src/linked_item.rs]: ah ok

**Pull request merged**